RecoVertex/BeamSpotProducer/plugins/alpaka/BeamSpotDeviceProducer.cc

```diff
@@ -0,0 +1,59 @@
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "DataFormats/BeamSpot/interface/BeamSpotHostProduct.h"
+#include "DataFormats/BeamSpot/interface/BeamSpotPOD.h"
+#include "DataFormats/BeamSpot/interface/alpaka/BeamSpotDeviceProduct.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EDPutToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/global/EDProducer.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class BeamSpotDeviceProducer : public global::EDProducer<> {
+  public:
+    BeamSpotDeviceProducer(edm::ParameterSet const& config)
+        : legacyToken_{consumes(config.getParameter<edm::InputTag>("src"))}, deviceToken_{produces()} {}
+
+    void produce(edm::StreamID, device::Event& event, device::EventSetup const& setup) const override {
+      reco::BeamSpot const& beamspot = event.get(legacyToken_);
+
+      BeamSpotHostProduct hostProduct{event.queue()};
+      hostProduct->x = beamspot.x0();
+      hostProduct->y = beamspot.y0();
+      hostProduct->z = beamspot.z0();
+      hostProduct->sigmaZ = beamspot.sigmaZ();
+      hostProduct->beamWidthX = beamspot.BeamWidthX();
+      hostProduct->beamWidthY = beamspot.BeamWidthY();
+      hostProduct->dxdz = beamspot.dxdz();
+      hostProduct->dydz = beamspot.dydz();
+      hostProduct->emittanceX = beamspot.emittanceX();
+      hostProduct->emittanceY = beamspot.emittanceY();
+      hostProduct->betaStar = beamspot.betaStar();
+
+      if constexpr (std::is_same_v<Device, alpaka::DevCpu>) {
```

Comment:
Is the point here to avoid the `memcpy()` for CPU backends? If that is the case, I think single instances of this pattern could be acceptable, but for wider use we should develop some abstraction (we have certainly discussed about it with @fwyzard, and I might have started to prototype something some time ago).

Reply:
Yes it is. Sincerely I don't remember why we don't rely on the automatic copy provided by the framework in this case. I would leave this to be checked and eventually corrected in a follow-up if this sounds reasonable to you.

Reply:
The framework automatic copy for an EDProducer is device-to-host, while here we need a host-to-device copy.

We are going to have a bunch of these (raw data, beam spot, PF input), but I'm not too concerned about abstracting it yet.
I think we can devise something later on, at the same time as we work on unified memory.

Reply:
Tracked here: https://github.com/cms-sw/cmssw/issues/43796#issuecomment-1912011782 .